app/src/main/java/org/sagebionetworks/research/mpower/medication/MedicationAdapter.kt

```diff
@@ -0,0 +1,193 @@
+import android.view.LayoutInflater
+import android.view.View
+import android.view.ViewGroup
+import android.widget.TextView
+import kotlinx.android.synthetic.main.medication_add.view.*
+import kotlinx.android.synthetic.main.medication_dosage.view.*
+import kotlinx.android.synthetic.main.medication_schedule.view.*
+import org.sagebionetworks.research.mpower.R.layout
+import org.sagebionetworks.research.mpower.medication.Type.ADD
+import org.sagebionetworks.research.mpower.medication.Type.DOSAGE
+import org.sagebionetworks.research.mpower.medication.Type.SCHEDULE
+import org.slf4j.LoggerFactory
+import java.text.SimpleDateFormat
+import java.util.Calendar
+
+class MedicationAdapter(var name: String, var items : List<MedicationItem>, val context: Context, val listener: Listener)
+    : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
+
+    private val LOGGER = LoggerFactory.getLogger(
+            MedicationAdapter::class.java)
+
+    override fun getItemCount(): Int {
+        return items.size
+    }
+
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
+        lateinit var holder: RecyclerView.ViewHolder
+        when(viewType) {
+            0 ->
+                holder = DosageViewHolder(
+                        LayoutInflater.from(context).inflate(layout.medication_dosage, parent, false))
+            1 ->
+                holder = ScheduleViewHolder(
+                        LayoutInflater.from(context).inflate(layout.medication_schedule, parent, false))
+            2 ->
+                holder = AddViewHolder(
+                        LayoutInflater.from(context).inflate(layout.medication_add, parent, false))
+        }
+        return holder
+    }
+
+    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
+        val item = items.get(position)
+        when(item.type) {
+            DOSAGE -> {
+                var dh = holder as DosageViewHolder
+                dh.bindView(item as Dosage)
+                dh.input.addTextChangedListener(object : TextWatcher {
+                    override fun beforeTextChanged(p0: CharSequence?, p1: Int, p2: Int, p3: Int) {}
+
+                    override fun onTextChanged(p0: CharSequence?, p1: Int, p2: Int, p3: Int) {
+                        var len = p0?.length
+                        when(len) {
+                            0 -> listener.enableNext(false)
+                            else -> listener.enableNext(true)
+                        }
+                    }
+
+                    override fun afterTextChanged(p0: Editable?) {}
+                })
+            }
+            SCHEDULE -> {
+                var sh = holder as ScheduleViewHolder
+                var schedule = item as Schedule
+                sh.bindView(schedule)
+                sh.checkbox.isChecked = schedule.anytime
+                sh.checkbox.setOnCheckedChangeListener{ buttonView, isChecked ->
+                    if(isChecked) {
```

Comment:
This code is equivalent to
sh.checkbox.setOnCheckedChangeListener { buttonView, isChecked ->
    schedule.anytime = isChecked
    listener.showAddSchedule(!isChecked)
    listener.setAnytime(schedule, isChecked)
}

val visibility = schedule.anytime ? View.GONE : View.VISIBLE
sh.dayContainer.visibility = visibility
sh.timeContainer.visibiliy = visiblity